File: leetcode/cpp/199.binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    vector<int> rightSideView(TreeNode *root)
    {
        if (!root)
            return {};
        queue<TreeNode *> q;
        vector<int> res, t;
        q.push(root);
        while (!q.empty())
        {
            size_t size = q.size();
            for (size_t i = 0; i < size - 1; ++i)
            {
                TreeNode *p = q.front();
                q.pop();
                if (p->left)
                    q.push(p->left);
                if (p->right)
                    q.push(p->right);
            }
            TreeNode *p = q.front();
            q.pop();
            res.push_back(p->val);

            if (p->left)
                q.push(p->left);
            if (p->right)
                q.push(p->right);
            // cout << q.size()<<endl;
        }
        return res;
    }
};

File: leetcode/python/199.binary-tree-right-side-view.py
from util import TreeNode
from typing import List

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        queue = [root]
        res = []
        while queue:
            size_ = len(queue)
            for i in range(size_):
                top = queue[i]
                if top.left:
                    queue.append(top.left)
                if top.right:
                    queue.append(top.right)
            res.append(queue[size_ - 1].val)
            queue = queue[size_:]
        return res
